src/Sulu/Component/HttpCache/EventSubscriber/InvalidationSubscriber.php

```diff
@@ -57,25 +93,122 @@ public function __construct(
     public static function getSubscribedEvents()
     {
         return [
-            Events::PUBLISH => ['invalidateDocumentForPublishing', -512],
+            Events::PUBLISH => ['invalidateDocumentBeforePublishing', 1024],
+            Events::UNPUBLISH => ['invalidateDocumentBeforeUnpublishing', 1024],
+            Events::REMOVE => ['invalidateDocumentBeforeRemoving', 1024],
         ];
     }
 
-    public function invalidateDocumentForPublishing(PublishEvent $event)
+    /**
+     * Invalidates all urls in the locale of the document of an already published document when it
+     * gets republished (eg on content change).
+     * This method is executed before the actual publishing of the document to avoid purging new urls.
+     *
+     * @param PublishEvent $event
+     */
+    public function invalidateDocumentBeforePublishing(PublishEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof ResourceSegmentBehavior || !$document->getPublished()) {
```

Comment:
For the second check you should check first if the document is an instance of `WorkflowStageBehavior`, don't you think?